Mathlib/Tactic/Set.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2022 Ian Benway. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Ian Benway.
+-/
+import Lean
+namespace Mathlib.Tactic
+open Lean Elab.Tactic Meta
+
+private def defineV (nm : Name) (tp : Expr) (tme : Expr) : TacticM Unit := do
+  liftMetaTactic1 fun mvarId => do
+    let h2 ← (define mvarId nm tp tme)
+    let (h, h2) ← intro1P h2
+    withMVarContext h2 do
+      return h2
+
+/--
+`set a := t with h` is a variant of `let a := t`. It adds the hypothesis `h : a = t` to
+the local context and replaces `t` with `a` everywhere it can.
+
+`set a := t with ←h` will add `h : t = a` instead.
+
+`set! a := t with h` does not do any replacing.
+
+```lean
+example (x : Nat) (h : x + x - x = 3) : x + x - x = 3 := by
+  set y := x with ←h2
+  sorry
+/-
+x : Nat
+y : Nat := x
+h : y + y - y = 3
+h2 : x = y
+⊢ y + y - y = 3
+-/
+```
+
+-/
+
+syntax (name := set) "set" ident (" : " term)? " := " term (" with " "←"? ident)? : tactic
+elab_rules : tactic
+|`(tactic| set $a:ident $[: $tp?:term]? := $pv:term $[with $[←%$rev?]? $h?:ident]?) => do
+  withMainContext do
+  match tp? with
+  | (some t) =>
```

Comment:
```suggestion
  | some t =>
```

Reply:
Changes pushed!

Reply:
Oops—I actually totally misunderstood that Simpa.lean file. The changes I pushed just simply compressed `set` and `set!` into one tactic, which is what Lean 3 does. I see that the Simpa file has a separate macro that distinguishes these two. Is there a benefit to that?

Reply:
If you go to the `test/Set.lean` file, you'll notice that `set!` is marked as an unknown tactic. In Lean 4, `foo!` is a fully formed identifier rather than a call to `foo` passing `!` as a parameter. That's why you'll need to tell Lean 4 what `set!` is.